File: src/main/java/com/ra/base_spring_boot/controller/GenreController.java
package com.ra.base_spring_boot.controller;

import com.ra.base_spring_boot.dto.ResponseWrapper;
import com.ra.base_spring_boot.dto.req.FormGenre;
import com.ra.base_spring_boot.dto.resp.GenreResponse;
import com.ra.base_spring_boot.model.Genre;
import com.ra.base_spring_boot.services.IGenreService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/v1/admin/")
@RequiredArgsConstructor
@Tag(name = "Genre Management", description = "APIs for managing genres (Admin only)")
public class GenreController {
    
    @Autowired
    private final IGenreService genreService;

    @Operation(summary = "Get all genres", description = "Retrieve a list of all genres")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successful retrieval",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "403", description = "Unauthorized access")
    })
    @SecurityRequirement(name = "bearerAuth")
    @GetMapping("/genres")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> getAllGenres() {
        List<Genre> genres = genreService.getAllGenres();
        List<GenreResponse> response = genres.stream().map(genre -> GenreResponse.builder()
                .id(genre.getId())
                .genreName(genre.getGenreName())
                .build()).collect(Collectors.toList());
        return ResponseEntity.status(HttpStatus.OK).body(
                ResponseWrapper.builder()
                        .status(HttpStatus.OK)
                        .code(200)
                        .data(response)
                        .build()
        );
    }

    @Operation(summary = "Add a new genre", description = "Create a new genre")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "201", description = "Genre created successfully",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "400", description = "Invalid input data"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access")
    })
    @SecurityRequirement(name = "bearerAuth")
    @PostMapping("/genre/create")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> addGenre(@Valid @RequestBody FormGenre genreRequestDto) {
        Genre genre = Genre.builder()
                .genreName(genreRequestDto.getGenreName())
                .build();
        Genre savedGenre = genreService.addGenre(genre);
        GenreResponse response = GenreResponse.builder()
                .id(savedGenre.getId())
                .genreName(savedGenre.getGenreName())
                .build();
        return ResponseEntity.status(HttpStatus.CREATED).body(
                ResponseWrapper.builder()
                        .status(HttpStatus.CREATED)
                        .code(201)
                        .data(response)
                        .build()
        );
    }

    @Operation(summary = "Update a genre", description = "Update an existing genre by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Genre updated successfully",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "400", description = "Invalid input data"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access"),
            @ApiResponse(responseCode = "404", description = "Genre not found")
    })
    @SecurityRequirement(name = "bearerAuth")
    @PutMapping("/genre/update/{id}")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> updateGenre(@Valid @PathVariable Long id, @RequestBody FormGenre genreRequestDto) {
        Genre genre = Genre.builder()
                .genreName(genreRequestDto.getGenreName())
                .build();
        Genre updatedGenre = genreService.updateGenre(id, genre);
        GenreResponse response = GenreResponse.builder()
                .id(updatedGenre.getId())
                .genreName(updatedGenre.getGenreName())
                .build();
        return ResponseEntity.status(HttpStatus.OK).body(
                ResponseWrapper.builder()
                        .status(HttpStatus.OK)
                        .code(200)
                        .data(response)
                        .build()
        );
    }

    @Operation(summary = "Delete a genre", description = "Delete a genre by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Genre deleted successfully",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "404", description = "Genre not found"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access")
    })
    @SecurityRequirement(name = "bearerAuth")
    @DeleteMapping("/genre/delete/{id}")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> deleteGenre(@PathVariable Long id) {
        genreService.deleteGenre(id);
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body(
                ResponseWrapper.builder()
                        .status(HttpStatus.NO_CONTENT)
                        .code(204)
                        .data("Genre deleted successfully")
                        .build()
        );
    }
}

File: src/controllers/admin/genreController.js
import genreService from '../../services/genreService.js';

class AdminGenreController {
  // Get all genres (Admin API)
  async getAllGenres(req, res) {
    try {
      const genres = await genreService.getAllGenres();
      
      // Convert to response format
      const response = genres.map(genre => ({
        id: genre.id,
        genreName: genre.name
      }));
      
      res.status(200).json({
        status: 200,
        code: 200,
        data: response
      });
    } catch (error) {
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Add new genre (Admin API)
  async addGenre(req, res) {
    try {
      const { genreName } = req.body;
      
      const genre = await genreService.addGenre({
        name: genreName
      });
      
      const response = {
        id: genre.id,
        genreName: genre.name
      };
      
      res.status(201).json({
        status: 201,
        code: 201,
        data: response
      });
    } catch (error) {
      if (error.message.includes('already exists')) {
        return res.status(409).json({
          status: 409,
          code: 409,
          message: error.message
        });
      }
      
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Update genre (Admin API)
  async updateGenre(req, res) {
    try {
      const { id } = req.params;
      const { genreName } = req.body;
      
      const genre = await genreService.updateGenre(id, {
        name: genreName
      });
      
      const response = {
        id: genre.id,
        genreName: genre.name
      };
      
      res.status(200).json({
        status: 200,
        code: 200,
        data: response
      });
    } catch (error) {
      if (error.message === 'Genre not found') {
        return res.status(404).json({
          status: 404,
          code: 404,
          message: 'Genre not found'
        });
      }
      
      if (error.message.includes('already exists')) {
        return res.status(409).json({
          status: 409,
          code: 409,
          message: error.message
        });
      }
      
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Delete genre (Admin API)
  async deleteGenre(req, res) {
    try {
      const { id } = req.params;
      
      await genreService.deleteGenre(id);
      
      res.status(204).json({
        status: 204,
        code: 204,
        message: 'Genre deleted successfully'
      });
    } catch (error) {
      if (error.message === 'Genre not found') {
        return res.status(404).json({
          status: 404,
          code: 404,
          message: 'Genre not found'
        });
      }
      
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }
}

export default new AdminGenreController();
